File: 003.前端代码（前端工程师）/frontend-react-java/src/lib/category-presentation.test.ts
// @vitest-environment jsdom
import { describe, it, expect } from 'vitest'
import { getCategoryPresentation, getCategoryPresentationById } from './category-presentation'
import type { Category as ApiCategory } from '../api/categories'

describe('getCategoryPresentation', () => {
  it('looks up known expense category by id', () => {
    const c: ApiCategory = {
      id: 'expense-餐饮',
      type: 'expense',
      name: '餐饮',
      icon: '🍔',
      color: '#4299E1',
      sortOrder: 0,
      isPreset: true,
    }
    const out = getCategoryPresentation(c)
    expect(out.icon).toBe('restaurant')
    expect(out.colorToken).toBe('cat-blue')
    expect(out.colorHex).toBe('#4299E1')
  })

  it('looks up known income category by id', () => {
    const c: ApiCategory = {
      id: 'income-工资',
      type: 'income',
      name: '工资',
      icon: '💰',
      color: '#10b981',
      sortOrder: 0,
      isPreset: true,
    }
    const out = getCategoryPresentation(c)
    expect(out.icon).toBe('payments')
    expect(out.colorToken).toBe('secondary')
    expect(out.colorHex).toBe('#10b981')
  })

  it('falls back to outline/more_horiz for unknown id with no icon/color', () => {
    // 后端 icon/color 缺失(老数据 / 删除一半) → 退回 outline 兜底
    const c: ApiCategory = {
      id: 'expense-mystery',
      type: 'expense',
      name: '神秘',
      icon: '',
      color: '',
      sortOrder: 99,
      isPreset: false,
    }
    const out = getCategoryPresentation(c)
    expect(out.icon).toBe('more_horiz')
    expect(out.colorToken).toBe('outline')
    expect(out.colorHex).toBe('#727782')
  })

  it('uses backend icon/color for unknown id (custom category)', () => {
    // 自定义分类:TABLE 查不到,直接用后端存的 c.icon (emoji) / c.color (#RRGGBB)
    const c: ApiCategory = {
      id: 'uuid-custom-1',
      type: 'expense',
      name: '测试自定义',
      icon: '🎯',
      color: '#A0AEC0',
      sortOrder: 99,
      isPreset: false,
    }
    const out = getCategoryPresentation(c)
    expect(out.icon).toBe('🎯')
    expect(out.colorHex).toBe('#A0AEC0')
  })

  it('covers all 9 expense + 5 income default categories', () => {
    const expenseIds = ['餐饮', '交通', '购物', '娱乐', '居住', '医疗', '教育', '通讯', '其他']
    const incomeIds = ['工资', '兼职', '理财', '红包', '其他']
    for (const n of expenseIds) {
      const out = getCategoryPresentationById(`expense-${n}`)
      expect(out.colorToken, `expense-${n}`).toBeTruthy()
    }
    for (const n of incomeIds) {
      const out = getCategoryPresentationById(`income-${n}`)
      expect(out.colorToken, `income-${n}`).toBeTruthy()
    }
  })
})

describe('getCategoryPresentationById', () => {
  it('returns same shape as getCategoryPresentation', () => {
    const a = getCategoryPresentationById('expense-购物')
    expect(a.icon).toBe('shopping_bag')
    expect(a.colorToken).toBe('cat-pink')
  })

  it('falls back consistently for unknown id', () => {
    expect(getCategoryPresentationById('xxx').icon).toBe('more_horiz')
  })
})

File: 007.跨端APP应用（移动端开发工程师）/flutter-project/test/unit/category_presentation_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:qingzhang_app/core/api/models.dart';
import 'package:qingzhang_app/core/utils/category_presentation.dart';

Category _cat(String id, String name, {String color = '#FF0000', String icon = ''}) {
  return Category(
    id: id,
    type: CategoryType.expense,
    name: name,
    icon: icon,
    color: color,
    sortOrder: 0,
    isPreset: false,
  );
}

void main() {
  test('preset expense 餐饮 → 后端 V2 emoji 🍜', () {
    // emoji/color 全部走后端,测试也得传后端值;不再从前端 _TABLE 取
    final p = presentCategory(_cat('preset-expense-餐饮', '餐饮', icon: '🍜', color: '#4299E1'));
    expect(p.icon, '🍜');
    expect(p.color, '#4299E1');
  });

  test('preset income 工资 → 后端 V2 emoji 💰', () {
    final p = presentCategory(_cat(
      'preset-income-工资',
      '工资',
      icon: '💰',
      color: '#10b981',
    ),);
    expect(p.icon, '💰');
    expect(p.color, '#10b981');
  });

  test('unknown id falls back to non-empty emoji', () {
    final p = presentCategory(_cat('custom-1', '我的类'));
    expect(p.icon, isNotEmpty);
  });

  test('empty icon → FALLBACK_ICON', () {
    final p = presentCategory(_cat('custom-2', '空图标', icon: '', color: '#abc'));
    expect(p.icon, '⋯');
  });

  test('empty color → FALLBACK_COLOR', () {
    final p = presentCategory(_cat('custom-3', '空颜色', icon: '🎯', color: ''));
    expect(p.color, '#727782');
  });

  test('MS ligature icon → FALLBACK_ICON(避免跨端显示字面文字)', () {
    final p = presentCategory(_cat('custom-4', '老 ligature', icon: 'restaurant', color: '#fff'));
    expect(p.icon, '⋯');
  });
}
